src/mscorlib/corefx/System/Runtime/InteropServices/StringBuffer.cs

```diff
@@ -283,44 +269,40 @@ public unsafe void CopyFrom(uint bufferIndex, string source, int sourceIndex = 0
             if (count < 0 || source.Length - count < sourceIndex) throw new ArgumentOutOfRangeException(nameof(count));
 
             if (count == 0) return;
-            uint lastIndex = bufferIndex + (uint)count;
+            int lastIndex = bufferIndex + (int)count;
             if (_length < lastIndex) Length = lastIndex;
 
-            fixed (char* content = source)
+            fixed (char* pSourceBuffer = source)
+            fixed (char* pDestinationBuffer = UnderlyingArray)
             {
                 Buffer.MemoryCopy(
-                    source: content + sourceIndex,
-                    destination: CharPointer + bufferIndex,
-                    destinationSizeInBytes: checked((long)(ByteCapacity - (bufferIndex * sizeof(char)))),
-                    sourceBytesToCopy: (long)count * sizeof(char));
+                    source: pSourceBuffer + sourceIndex,
+                    destination: pDestinationBuffer + bufferIndex,
+                    destinationSizeInBytes: checked((long)((Capacity - bufferIndex) * sizeof(char))),
+                    sourceBytesToCopy: checked((long)count * sizeof(char)));
             }
         }
 
         /// <summary>
         /// Trim the specified values from the end of the buffer. If nothing is specified, nothing is trimmed.
         /// </summary>
-        public unsafe void TrimEnd(char[] values)
+        public void TrimEnd(char[] values)
         {
             if (values == null || values.Length == 0 || _length == 0) return;
 
-            char* end = CharPointer + _length - 1;
-
-            while (_length > 0 && Array.IndexOf(values, *end) >= 0)
+            while (_length > 0 && Array.IndexOf(values, _buffer[_length - 1]) >= 0)
             {
                 Length = _length - 1;
-                end--;
             }
         }
 
         /// <summary>
         /// String representation of the entire buffer. If the buffer is larger than the maximum size string (int.MaxValue) this will throw.
         /// </summary>
         /// <exception cref="InvalidOperationException">Thrown if the buffer is too big to fit into a string.</exception>
-        public unsafe override string ToString()
+        public override string ToString()
         {
-            if (_length == 0) return string.Empty;
-            if (_length > int.MaxValue) throw new InvalidOperationException();
-            return new string(CharPointer, startIndex: 0, length: (int)_length);
+            return new string(_buffer, startIndex: 0, length: _length);
```

Comment:
The special casing for zero length is done by all string constructors already. No need to replicate it here. It would be worth replicating only if zero length is the hot path and the extra code is worth it to avoid the extra call.